### PR TITLE
let image_url compute bounds

### LINKED ISSUE #3187: plotting/file/image_url.py doesn't work without setting {x,y}_range
Fails with "could not set initial ranges" from `common/plot.coffee`.

### PR DESCRIPTION
**Pull request by bryevdv**
- [x] issues: fixes #3187
- [x] release document entry (if new feature or API change)

### CODE REVIEW

**Comment by bryevdv** [comment]
Updated example with data ranges:

screen shot 2018-09-20 at 18 15 53

File: bokehjs/src/lib/models/glyphs/image_url.ts [hunk 1]
```diff
@@ -16,6 +16,7 @@ export interface ImageURLData extends XYGlyphData {
   _angle: Arrayable<number>
   _w: Arrayable<number>
   _h: Arrayable<number>
+  _bounds_rect : Rect
```
Comment (mattpap) [inline bokehjs/src/lib/models/glyphs/image_url.ts]: whitespace

File: bokehjs/src/lib/models/glyphs/image_url.ts [hunk 2]
```diff
@@ -74,6 +75,32 @@ export class ImageURLView extends XYGlyphView {
       }
       img.src = this._url[i]
     }
+
+    const xs : Array<number> = []
```
Comment (mattpap) [inline bokehjs/src/lib/models/glyphs/image_url.ts]: For consistency across code base, the type should be `number[]`. This is actually enforced by a tslint rule, though tslint isn't enabled on travis ci (this is fixed in PR #8085).

Comment (mattpap) [reply bokehjs/src/lib/models/glyphs/image_url.ts]: Also, whitespace.

File: bokehjs/src/lib/models/glyphs/image_url.ts [hunk 3]
```diff
@@ -74,6 +75,32 @@ export class ImageURLView extends XYGlyphView {
...
+    const ys : Array<number> = []
+    for (var i=0; i < this._x.length; i++) {
+      xs.push(this._x[i])
+      ys.push(this._y[i])
+    }
+
+    // if the width/height are in screen units, don't try to include them in bounds
+    if (this.model.properties.w.units == "data") {
+      for (var i=0; i < this._x.length; i++) {
```
Comment (mattpap) [inline bokehjs/src/lib/models/glyphs/image_url.ts]: `var` isn't allowed across the code base. `let` should be used instead. This is enforced by tslint (see above).

Comment (mattpap) [reply bokehjs/src/lib/models/glyphs/image_url.ts]: Also, whitespace.

File: bokehjs/src/lib/models/glyphs/image_url.ts [hunk 4]
```diff
@@ -74,6 +75,32 @@ export class ImageURLView extends XYGlyphView {
...
+        xs.push(this._x[i]+this._w[i])
+      }
+    }
+    if (this.model.properties.h.units == "data") {
+      for (var i=0; i < this._x.length; i++) {
+        ys.push(this._y[i]+this._h[i])
+      }
+    }
+
+    const minX = Math.min(...xs)
```
Comment (mattpap) [inline bokehjs/src/lib/models/glyphs/image_url.ts]: The number of arguments allowed to a function is platform/browser dependent. `min()` and `max()` from `core/util/arrayable` should be used instead.

File: bokehjs/src/lib/models/glyphs/image_url.ts [hunk 5]
```diff
@@ -74,6 +75,32 @@ export class ImageURLView extends XYGlyphView {
...
+    const maxX = Math.max(...xs)
+    const minY = Math.min(...ys)
+    const maxY = Math.max(...ys)
+
+    this._bounds_rect =  {minX, maxX, minY, maxY}
```
Comment (mattpap) [inline bokehjs/src/lib/models/glyphs/image_url.ts]: Whitespace as well.

**Comment by mattpap** [comment]
I fixed those in https://github.com/bokeh/bokeh/pull/8085/commits/e2cd9ba123d08bce8e4eea6d348893343eab075b, to get tests passing in PR #8085.

**Pull request merged**